src/network/core/os_abstraction.cpp

```diff
@@ -0,0 +1,184 @@
+ * @return True iff the operation would block.
+ */
+bool NetworkError::WouldBlock() const
+{
+#if defined(_WIN32)
+	return this->error == WSAEWOULDBLOCK;
+#elif EWOULDBLOCK != EAGAIN
+	/* Usually EWOULDBLOCK and EAGAIN are the same, but sometimes they are not
+	 * and the POSIX.1 specification states that either should be checked. */
+	return this->error == EWOULDBLOCK || this->error == EAGAIN;
+#else
+	return this->error == EWOULDBLOCK;
+#endif
+}
+
+/**
+ * Check whether this error describes a connection reset.
+ * @return True iff the connection is reset.
+ */
+bool NetworkError::IsConnectionReset() const
+{
+#if defined(_WIN32)
+	return this->error == WSAECONNRESET;
+#else
+	return this->error == ECONNRESET;
+#endif
+}
+
+/**
+ * Check whether this error describes a connect is in progress.
+ * @return True iff the connect is already in progress.
+ */
+bool NetworkError::IsConnectInProgress() const
+{
+#if defined(_WIN32)
+	return this->error == WSAEWOULDBLOCK;
+#else
+	return this->error == EINPROGRESS;
+#endif
+}
+
+/**
+ * Get the C-string representation of the error message.
+ * @return The C-string representation that will get overwritten by next calls.
+ */
+const char *NetworkError::AsString() const
+{
+#if defined(_WIN32)
+	static char buffer[512];
+	if (FormatMessageA(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS, NULL, this->error,
+		MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), buffer, sizeof(buffer), NULL) == 0) {
+		seprintf(buffer, lastof(buffer), "Unknown error %d", this->error);
+	}
+	return buffer;
+#else
+	return strerror(this->error);
+#endif
+}
+
+/**
+ * Check whether an error was actually set.
+ * @return True iff an error was set.
+ */
+bool NetworkError::HasError() const
+{
+	return this->error != 0;
+}
+
+/**
+ * Get the last network error.
+ * @return The network error.
+ */
+/* static */ NetworkError NetworkError::GetLast()
+{
+	return NetworkError();
+}
+
+/**
+ * Convenience methods to the get C-string represenation of the last error.
+ * @return The C-string representation that will get overwritten by next calls.
+ */
+/* static */ const char *NetworkError::GetLastAsString()
+{
+	return GetLast().AsString();
+}
+
+
+/**
+ * Try to set the socket into non-blocking mode.
+ * @param d The socket to set the non-blocking more for.
+ * @return True if setting the non-blocking mode succeeded, otherwise false.
+ */
+bool SetNonBlocking(SOCKET d)
+{
+#if defined(_WIN32)
+	u_long nonblocking = 1;
+	return ioctlsocket(d, FIONBIO, &nonblocking) == 0;
+#elif defined __EMSCRIPTEN__
+	return true;
+#else
+	int nonblocking = 1;
+	return ioctl(d, FIONBIO, &nonblocking) == 0;
+#endif
+}
+
+/**
+ * Try to set the socket to not delay sending.
+ * @param d The socket to disable the delaying for.
+ * @return True if disabling the delaying succeeded, otherwise false.
+ */
+bool SetNoDelay(SOCKET d)
+{
+#ifdef __EMSCRIPTEN__
+	return true;
+#else
+	/* XXX should this be done at all? */
```

Comment:
Maybe also time to part from this comment. No clue what it refers to, and there is nothing here that can be removed :P

I know, not your code, but while at it ... :D

(bonus points if you add a space between `char` and `*` on the lines below :D)